### DISCUSSION TITLE
Question about "input.apply_p_jones" and "output.apply_p_jones_inv" functionality

### DISCUSSION THREAD

**Discussion by AKHughes1994** [ratt-ru/QuartiCal#328] [Q&A]
Hi, 

I have a question about the terminology of the title parameters. Namely the help text, 

`input_model.apply_p_jones
    Determines whether P-Jones (parallactic angle rotation) is applied to the model. This affects both measurement set columns and predicted components. Care
    must taken when using this option and output.apply_p_jones_inv. Default: False
`

If I understand this correctly, is this saying that it will apply the parallactic angle rotation to the model AND the input data column visibilities? 

For a fuller context, I am trying to perform full-polarization self-calibration. 

I have a DATA column that has been split from a larger ms file and has all the Gain terms applied (i.e., K, Gp, Ga, BP, Df, Xf) with parallactic angle correction applied (This is done in casa). I then image the DATA column, using the IQUV model to population MODEL. From here, if I want to self-calibrate. 

Would I set `input_model.apply_p_jones = True` to re-rotate according to the parallactic angle? Then, also set  `output.apply_p_jones_inv = True` to re-apply the parallactic angle correction to the now self-calibrated corrected visibilities? 

Am I on the right track here, or have I misunderstood the purpose of these flags? 

Thanks for the help,
Andrew

**Comment by JSKenyon** [comment]
Hi Andrew! 

> If I understand this correctly, is this saying that it will apply the parallactic angle rotation to the model AND the input data column visibilities?

I think I can see why this may be misleading. This is actually referring to the case where you have both columns and skymodels present in the model e.g. `input_model.recipe="MODEL_DATA~bright_source.lsm.html:bright_source.lsm.html"`. The parallactic angle will be applied to all the components of the model (the MODEL_DATA column and bright_source.lsm.html component model in the above example). It will not be applied to the DATA column.

> Would I set `input_model.apply_p_jones = True` to re-rotate according to the parallactic angle? Then, also set `output.apply_p_jones_inv = True` to re-apply the parallactic angle correction to the now self-calibrated corrected visibilities?

This can get a little complicated so I will try and explain the logic and then you can decide what applies to your use case. The uncalibrated data is in the telescope frame but our model will typically be in the sky frame. Applying parallactic angle to the model moves it into the telescope frame (where we want to do our calibration). This corresponds to `input_model.apply_p_jones`. Once we have done the calibration and are forming up corrected data, we want to move the corrected data into the sky frame by applying the inverse of the parallactic angle matrix. This corresponds to `output.apply_p_jones_inv`. 

I believe that you are on the right track, but hopefully the above makes it a little clearer. Don't hesitate to ask if you need further clarification.

**Comment by AKHughes1994** [comment]
Hi, 

Ok, that makes sense. 

For more details about this use case. The current approach is to use CASA to perform the standard "1GC" calibration (i.e., K, Gp, Ga, BP, Df, Xf). But, to my knowledge, CASA cannot explicitly solve for off-diagonal terms when self-calibrating and thus isn't the best for polarized self-calibration. I've had much success using Quartical for polarization self-calibration of VLA data, where I performed 1GC with Casa and split out the CORRECTED_DATA. into a new MS file, then self-calibrated the new MS file with the model produced from initial imaging. 

But, from what I understand, if I split out CORRECTED_DATA with a parallactic angle correction applied (parang=True in CASA) for a linear-feed instrument (like MeerKAT), I need to be very careful because the P-matrix doesn't commute, as it does for circular feeds. Therefore, there is a risk of "over-correcting" parallactic angle rotation (corrupting EVPA measurements) when performing self-calibration with Quartical on CASA-calibrated 1GC data.

I'd imagine I could: 
1. Split out the CASA-corrected data without parallactic angle corrections (parang=False)
2. Image the split data to build a model without parallactic angle corrections
3. Run Quartical self-cal with `input_model.apply_p_jones = False` and `output.apply_p_jones_inv = True`  since I built the model in the telescope frame but want the output visibilities to be in the sky frame. 
4. Reap the rewards

But I'm not 100% sure. 

If I didn't screw up the explanation (which may have occurred), is there a recommended use case for a scenario like this? Obviously, this is hybridizing CASA and Quartical, and the answer may be to convert to Quartical fully, so I apologize for bringing external software into the problem; I appreciate any help/advice.

Andrew

**Reply by JSKenyon** [reply]
> For more details about this use case. The current approach is to use CASA to perform the standard "1GC" calibration (i.e., K, Gp, Ga, BP, Df, Xf). But, to my knowledge, CASA cannot explicitly solve for off-diagonal terms when self-calibrating and thus isn't the best for polarized self-calibration. I've had much success using Quartical for polarization self-calibration of VLA data, where I performed 1GC with Casa and split out the CORRECTED_DATA. into a new MS file, then self-calibrated the new MS file with the model produced from initial imaging.

Glad to hear that you are having some success doing this with QuartiCal! 

> But, from what I understand, if I split out CORRECTED_DATA with a parallactic angle correction applied (parang=True in CASA) for a linear-feed instrument (like MeerKAT), I need to be very careful because the P-matrix doesn't commute, as it does for circular feeds. Therefore, there is a risk of "over-correcting" parallactic angle rotation (corrupting EVPA measurements) when performing self-calibration with Quartical on CASA-calibrated 1GC data.

Agreed. Ideally, you never want to calibrate on DATA/CORRECTED_DATA with the parallactic angle correction applied. This may make things a little complicated. I think one would need to form up MODEL_DATA from CORRECTED_DATA with parang=True in casa, then split out a version with parang=False for calibration with QC.  

> I'd imagine I could:
> 
> 1. Split out the CASA-corrected data without parallactic angle corrections (parang=False)
> 2. Image the split data to build a model without parallactic angle corrections
> 3. Run Quartical self-cal with `input_model.apply_p_jones = False` and `output.apply_p_jones_inv = True`  since I built the model in the telescope frame but want the output visibilities to be in the sky frame.
> 4. Reap the rewards

As mentioned above, I think that it is a little more complicated as you don't want to build the model in the telescope frame. I think my strategy would be something like this:

1.  Split out the CASA-corrected data with parallactic angle corrections (parang=True) i.e. CORRECTED_DATA will have the inverse of all the CASA gain terms and the parallactic angle applied.
2. Populate MODEL_DATA (or a component model) using the result from 1. This model should be in the sky frame.
3. Split out the CASA-corrected data without parallactic angle corrections (parang=False) i.e. CORRECTED_DATA will have the inverse of all the casa gain terms applied but not the parallactic angle.
4. Using the model from 2 and the data from 3, run Quartical self-cal with `input_model.apply_p_jones = True` and `output.apply_p_jones_inv = True` i.e. take the model into the telescope frame for calibration and move the data into the sky frame after correction.
5. Profit?

> If I didn't screw up the explanation (which may have occurred), is there a recommended use case for a scenario like this? Obviously, this is hybridizing CASA and Quartical, and the answer may be to convert to Quartical fully, so I apologize for bringing external software into the problem; I appreciate any help/advice.

I don't think we have a tried and tested approach although I know that both QuartiCal and its predecessor CubiCal are regularly used after CASA 1GC. No need to apologize. That said, the intention is that QuartiCal should be able to handle the 1GC calibration too. If you wanted to give that a go at some point in the future, I would be happy to help (and inevitably fix the things which don't work).

**Reply by AKHughes1994** [reply]
Thanks for this; I agree your method would make sense, but there is a caveat. 

It looks like if casacore has an outdated LEAP table the p_jones terms cause QC to fail,  

![image](https://github.com/ratt-ru/QuartiCal/assets/49698839/c0f45e4f-74b7-44b9-84d6-8551e863b5cd)

Still trying to figure out how to get around that. Any suggestions? 

For context, I am just using the default version that gets installed using pip (v0.2.2)

**Reply by JSKenyon** [reply]
Could you please try `pip install -U casadata` inside your virtual environment and let me know if it resolves the problem? It occurs to me that this may just be a missing dependency in QuartiCal - there was a time when this had to be done manually. If it does work I will make sure it is fixed in the next release.

**Reply by JSKenyon** [reply]
@AKHughes1994 Apologies for the ping, but did the above fix the issue?

**Reply by AKHughes1994** [reply]
Hi @JSKenyon, sorry, I'll try this today/tommorow and see if it resolves the issue! I haven't had time to get around to it yet.

**Reply by AKHughes1994** [reply]
@JSKenyon, it looks like it didn't fix the issue. Here is some extra information, 

1. I have installed casa data (v2024.01.15) 
![image](https://github.com/ratt-ru/QuartiCal/assets/49698839/44bd15da-37f5-4ed3-89fb-ab6e54379c25)
But the error is the same. In particular, it only seems to search in `/usr/share/casacore`, so QC might not be looking in casadata  within the Venv. for 

2. My observations are from 2024-01-30, so if the version name is a date, I wonder if the pip installation for casa data will be sufficiently current. I manually opened the "MJD" column of the "IERSeop97" table in the casadata and the maximum date is only 2023-11-15, exactly two months behind the release. Let me caveat that I don't fully know how casa does these calculations, so that date may not correspond to the most recent observations where these corrections can be applied. 

Andrew

**Reply by AKHughes1994** [reply]
Moreover, 

If I run `!update-data` inside a casa  instance, it gets an updated version from the pip release: 

![image](https://github.com/ratt-ru/QuartiCal/assets/49698839/09ce26b8-afe7-4214-828b-cdb445a1f6ed)

However, this only goes up to 2023-12-31, so I might be S.O.L, given the recency of my observations. 

Andrew

**Reply by JSKenyon** [reply]
That is a bit strange - I believe I can reproduce your problem by renaming my .casarc file:
```
2024-02-28 07:44:04     WARN    MeasIERS::findTab (file /code/measures/Measures/MeasIERS.cc, line 389)  Requested data table IERSeop97 cannot be found in the searched directories:
2024-02-28 07:44:04     WARN    MeasIERS::findTab (file /code/measures/Measures/MeasIERS.cc, line 389)+ /usr/share/casacore/data/ephemerides/
2024-02-28 07:44:04     WARN    MeasIERS::findTab (file /code/measures/Measures/MeasIERS.cc, line 389)+ /usr/share/casacore/data/geodetic/
2024-02-28 07:44:04     WARN    MeasIERS::findTab (file /code/measures/Measures/MeasIERS.cc, line 389)  Requested data table IERSpredict cannot be found in the searched directories:
2024-02-28 07:44:04     WARN    MeasIERS::findTab (file /code/measures/Measures/MeasIERS.cc, line 389)+ /usr/share/casacore/data/ephemerides/
2024-02-28 07:44:04     WARN    MeasIERS::findTab (file /code/measures/Measures/MeasIERS.cc, line 389)+ /usr/share/casacore/data/geodetic/
2024-02-28 07:44:04     INFO    MeasIERS::fillMeas(MeasIERS::Files, Double)     Requested JD 56529.5 is outside the range of the IERS (Earth axis data) table.
2024-02-28 07:44:04     INFO    MeasIERS::fillMeas(MeasIERS::Files, Double) +   Calculations will proceed with less precision
2024-02-28 07:44:04     WARN    MeasIERS::findTab (file /code/measures/Measures/MeasIERS.cc, line 389)  Requested data table TAI_UTC cannot be found in the searched directories:
2024-02-28 07:44:04     WARN    MeasIERS::findTab (file /code/measures/Measures/MeasIERS.cc, line 389)+ /usr/share/casacore/data/ephemerides/
2024-02-28 07:44:04     WARN    MeasIERS::findTab (file /code/measures/Measures/MeasIERS.cc, line 389)+ /usr/share/casacore/data/geodetic/
2024-02-28 07:44:04     SEVERE  MeasTable::dUTC(Double) (file /code/measures/Measures/MeasTable.cc, line 4283)  Cannot read leap second table TAI_UTC
```
Can you please give [Ben's solution](https://github.com/ratt-ru/QuartiCal/discussions/328#discussioncomment-8564834) a try? I think that you just need .casarc to point at something appropriate. 

Edit: Tried this myself and made .casarc point to `/home/jonathan/venvs/tmpenv/lib/python3.10/site-packages/casadata/__data__/`. That works but I think that Ben's solution will likely give you a more current, smaller and less environment dependent solution. I should catch this error and give the user some sensible advice.

**Reply by JSKenyon** [reply]
In case it was unclear, `.casarc` is a file that should live in your home (you may need to create it if it doesn't already exist). The contents of the file should include:
```
measures.directory: /path/to/casadata
```

**Reply by JSKenyon** [reply]
See #330.

**Comment by landmanbester** [comment]
My 2 cents on the matter is that you may get away with ignoring parallactic angle rotation during selfcal if you have corrected for it during 1GC (I think getting the order of the terms wrong at that stage will end up being a second or higher order effect). But you may also be able to do the full 1GC chain with QuartiCal now (I'm not sure what Df and Xf refer to above) thereby avoiding the need to calibrate corrected data, which is almost surely always dodgy

**Reply by AKHughes1994** [reply]
For context Df = Leakage, and Xf = cross-hand phase

**Comment by bennahugo** [comment]
If you are running Ubuntu you can install casacore-data and set .casarc
accordingly (as you would need to do for CASA installations).

Alternatively you can grab ephemeris directly from the NRAO:
mkdir -p /usr/share/casacore/data/
/usr/share/casacore/data/
rsync -avz rsync://casa-rsync.nrao.edu/casa-data .



On Thu, Feb 22, 2024 at 10:46 PM AKHughes1994 ***@***.***>
wrote:

> Thanks for this; I agree your method would make sense, but there is a
> caveat.
>
> It looks like if casacore has an outdated LEAP table the p_jones terms
> cause QC to fail,
>
> image.png (view on web)
> <https://github.com/ratt-ru/QuartiCal/assets/49698839/c0f45e4f-74b7-44b9-84d6-8551e863b5cd>
>
> Still trying to figure out how to get around that. Any suggestions?
>
> For context, I am just using the default version that gets installed using
> pip (v0.2.2)
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ratt-ru/QuartiCal/discussions/328#discussioncomment-8561528>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AB4RE6RJGN2LRHPME4LVX2TYU6VCTAVCNFSM6AAAAABDRZNJICVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM4DKNRRGUZDQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
--
Benjamin Hugo

**Reply by JSKenyon** [reply]
This is also a valid option, but please try the `pip install -U casadata` option first, as that can be built into QuartiCal if it works. Thanks for the input Ben!

**Reply by sjperkins** [reply]
Just FYI, the casadata wheel is ~350MB and uncompresses to 833MB

```bash
~/venv/lib/python3.10/site-packages/casadata/__data__$ du -hd 1
80K	./gui
415M	./ephemerides
25M	./nrao
9.1M	./geodetic
45M	./demo
40M	./catalogs
301M	./alma
833M	.
```

by contrast `ftp://ftp.astron.nl/outgoing/Measures/WSRT_Measures.ztar` seems to be ~20MB

```bash
$ wget ftp://ftp.astron.nl/outgoing/Measures/WSRT_Measures.ztar
--2024-02-23 09:22:38--  ftp://ftp.astron.nl/outgoing/Measures/WSRT_Measures.ztar
           => ‘WSRT_Measures.ztar’
Resolving ftp.astron.nl (ftp.astron.nl)... 192.87.1.38
Connecting to ftp.astron.nl (ftp.astron.nl)|192.87.1.38|:21... connected.
Logging in as anonymous ... Logged in!
==> SYST ... done.    ==> PWD ... done.
==> TYPE I ... done.  ==> CWD (1) /outgoing/Measures ... done.
==> SIZE WSRT_Measures.ztar ... 20488360
==> PASV ... done.    ==> RETR WSRT_Measures.ztar ... done.
Length: 20488360 (20M) (unauthoritative)

WSRT_Measures.ztar                                                                                         100%[=====================================================================================================================================================================================================================================================================================>]  19.54M  5.52MB/s    in 4.3s    

2024-02-23 09:22:45 (4.52 MB/s) - ‘WSRT_Measures.ztar’ saved [20488360]
```

**Reply by JSKenyon** [reply]
It seems that there is substantially more stuff in the `casadata` wheel. That is not necessarily a good thing, but it is definitely more convenient than having to do it by hand/add extra steps to the installation. Perhaps it is worth investigating whether there is interest in a `casadata-minimal` package. For reference, `WSRT_Measures.ztar` only includes the `ephemerides` and `geodetic` folders from the wheel (approximately).

**Reply by sjperkins** [reply]
I'm intending to auto download the smaller measures in arcae at some point.

**Reply by AKHughes1994** [reply]
As a follow-up, I  grabbed ephemeris directly using the method from @bennahugo, and it solved my problem! 

Sorry for the delayed reply and Thanks!

**Reply by JSKenyon** [reply]
Glad that it fixed the problem! I am still thinking about the best way to handle this in general without requiring additional setup from users.
